server/src/weaverbird/backends/pypika_translator/translators/base.py

```diff
@@ -667,6 +669,75 @@ def ifthenelse(
 
         return StepContext(query, columns + [step.new_column])
 
+    @staticmethod
+    def _get_join_type(join_type: Literal['left', 'inner', 'left outer']) -> JoinType:
+        match join_type:
+            case 'left':
+                return JoinType.left
+            case 'left outer':
+                return JoinType.left_outer
+            case 'inner':
+                return JoinType.inner
+
+    @staticmethod
+    def _field_list_to_name_list(fields: list[Field]) -> list[str]:
+        return [field.alias or field.name for field in fields]
+
+    def join(
+        self: Self,
+        *,
+        builder: 'QueryBuilder',
+        prev_step_name: str,
+        columns: list[str],
+        step: "JoinStep",
+    ) -> StepContext:
+        try:
+            steps = Pipeline(steps=step.right_pipeline).steps
+        except Exception as exc:
+            raise NotImplementedError(
+                f"join is only possible with another pipeline: {exc}"
+            ) from exc
+        right_builder_ctx = self.__class__(
+            tables_columns=self._tables_columns, db_schema=self._db_schema_name
+        ).get_query_builder(steps=steps, query_builder=builder)
+        left_table = Table(prev_step_name)
+        right_table = Table(right_builder_ctx.table_name)
+
+        left_cols = [Field(col, table=left_table) for col in columns]
+        # Simply checking if a column is in left columns and right columns is not enough to
+        # determine if it needs to be aliased. In case of nested joins where a column is used in
+        # three or more datasets (not unlikely with columns like "id" or "name"), the aliased column
+        # may need to be suffixed several times. The join/inner_pypika_nested.yaml fixture shows an
+        # example of this kind of situation
+        right_cols: list[Field] = []
+        for col in right_builder_ctx.columns:
+            if col not in (all_cols := self._field_list_to_name_list(left_cols + right_cols)):
+                right_cols.append(Field(col, table=right_table))
+                continue
+            alias = col
+            while (alias := f'{alias}_right') in all_cols:
+                pass
+            right_cols.append(Field(col, table=right_table, alias=alias))
+
+        query = (
+            self.QUERY_CLS.from_(left_table)
+            .select(*left_cols, *right_cols)
+            .join(right_table, self._get_join_type(step.type))
+            .on(
+                Criterion.all(
+                    Field(f[0], table=left_table) == Field(f[1], table=right_table) for f in step.on
+                )
+            )
+            # Order of results is not consistent depending on the SQL Engine (inconsistencies
+            # observed with Athena and BigQuery).
+            .orderby(*(c[0] for c in step.on))
```

Comment:
Should we document this in the weaverbird's doc?

Reply:
Done in https://github.com/ToucanToco/weaverbird/pull/1407/commits/ee77a6b88ec559ab0df1a928728d6012bd6eb4af . I think all the docs deserve a refresh though :sweat_smile: